src/seq/coin_flipper.rs

```diff
@@ -0,0 +1,110 @@
+use crate::RngCore;
+
+pub(crate) struct CoinFlipper<R: RngCore> {
+    pub rng: R,
+    chunk: u32,
+    chunk_remaining: u32,
+}
+
+impl<R: RngCore> CoinFlipper<R> {
+    pub fn new(rng: R) -> Self {
+        Self {
+            rng,
+            chunk: 0,
+            chunk_remaining: 0,
+        }
+    }
+
+    #[inline]
+    /// Returns true with a probability of numerator / denominator
+    /// Uses an expected two bits of randomness
+    pub fn gen_ratio(&mut self, mut numerator: usize, denominator: usize) -> bool {
+        // Explanation:
+        // We are trying to return true with a probability of n / d
+        // If n >= d, we can just return true
+        // Otherwise there are two possibilities 2n < d and 2n >= d
+        // In either case we flip a coin.
+        // If 2n < d
+        //  If it comes up tails, return false
+        //  If it comes up heads, double n and start again
+        //  This is fair because (0.5 * 0) + (0.5 * 2n / d) = n / d and 2n is less than d (if 2n was greater than d we would effectively round it down to 1 by returning true)
+        // If 2n >= d
+        //   If it comes up tails, set n to 2n - d
+        //   If it comes up heads, return true
+        //   This is fair because (0.5 * 1) + (0.5 * (2n - d) / d) = n / d
+
+        while numerator < denominator {
+            //The exponent is the number of heads that need to be won in a row to not return false
+            let exponent = (numerator.leading_zeros() - denominator.leading_zeros()).min(32);
+
+            if exponent > 1 {
+                //n * 2^exponent < d
+                if self.flip_until_tails(exponent) {
+                    // all heads
+                    numerator <<= exponent;
+                } else {
+                    // a tails somewhere
+                    return false;
+                }
+            }
+            //Otherwise n * 2 is either greater than or very close to d
```

Comment:
E.g. `2n = 2*b011 = 6`, `d = b100 = 8`. What does "very close" mean? Either way, you're deviating from the stated algorithm above.